sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala

```diff
@@ -450,3 +454,85 @@ case class ScalaUDAF(
 
   override def nodeName: String = udaf.getClass.getSimpleName
 }
+
+case class ScalaAggregator[IN, BUF, OUT](
+    children: Seq[Expression],
+    agg: Aggregator[IN, BUF, OUT],
+    inputEncoderNR: ExpressionEncoder[IN],
+    isNullable: Boolean = true,
+    isDeterministic: Boolean = true,
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0)
+  extends TypedImperativeAggregate[BUF]
+  with NonSQLExpression
+  with UserDefinedExpression
+  with ImplicitCastInputTypes
+  with Logging {
+
+  private[this] lazy val inputEncoder = inputEncoderNR.resolveAndBind()
```

Comment:
we should resolve with `childrenSchema.toAttributes`

Reply:
ah wait, just realized we miss an important feature: type check. If the aggregator accepts int input, but the actual input column is string, what shall we do?

Looking at `ScalaUDAF`, we implement `ImplicitCastInputTypes` and define the expected input types. I think we should do the same thing for `ScalaAggregator`, and add tests for it.

Reply:
It already implements `ImplicitCastInputTypes` but I'll look at `childrenSchema.toAttributes` - maybe that would also help with the projection?

Reply:
Resolving with `childrenSchema.toAttributes` actually also caused several unit tests to fail. Combining this with bypassing the `inputProjection` also failed testing. My reading of the unit testing failures is that this change was breaking the `ImplicitCastInputTypes` functionality and/or aggregation on expressions. My take is we should leave it as-is, at least for the purposes of this PR.